File: p054.py
import sys, time
from itertools import *

CARD_VALUES = list("23456789TJQKA")
HAND_RANKS = [
    "HighCard",
    "OnePair",
    "TwoPair",
    "ThreeKind",
    "Straight",
    "Flush",
    "FullHouse",
    "FourKind",
    "StraightFlush"]


class Hand:
    def __init__(self, cards):
        self.cards = cards
        self.values = [CARD_VALUES.index(c[0])+1 for c in cards]
        self.suits = [c[1] for c in cards]

    @property
    def rank(self):
        distinct = len(set(self.values))
        longest = max(len(list(g)) 
                      for k, g in groupby(sorted(self.values)))
        if distinct == 2:
            rank = "FourKind" if longest == 4 else "FullHouse"
        elif distinct == 3:
            rank = "ThreeKind" if longest == 3 else "TwoPair"
        elif distinct == 4:
            rank = "OnePair"
        else:
            flush = len(set(self.suits)) == 1
            straight = max(self.values) - min(self.values) == 4
            rank = ("StraightFlush" if straight and flush
                    else "Straight" if straight 
                    else "Flush" if flush
                    else "HighCard")
        return HAND_RANKS.index(rank)
 
    def trumps(self, other):
        def tiebreakers(values):
            def group_score(value, group):
                return (value * (len(CARD_VALUES)+1) ** (len(list(group))-1))

            return reversed(sorted(starmap(group_score,
                                           groupby(sorted(values)))))
        if self.rank == other.rank:
            selfTiebreaker, otherTiebreaker = next(dropwhile(
                lambda tup: tup[0] == tup[1],
                zip(tiebreakers(self.values), tiebreakers(other.values))
            ))
            return selfTiebreaker > otherTiebreaker
        else:
            return self.rank > other.rank


if __name__ == "__main__":
    filename = sys.argv[1] if len(sys.argv) > 1 else "p054.txt"

    t0 = int(time.time()*1000)
    ans = list(starmap(lambda left, right: left.trumps(right),
                       map(lambda cards: (Hand(cards[0:5]), Hand(cards[5:10])),
                           (line.split() for line in open(filename))))).count(1)
    t1 = int(time.time()*1000)

    print(t1 - t0)

    print(ans)

File: p054.scala
import Function.tupled

object p054 {
  class Hand (val cards: Array[String]) extends Ordered[Hand] {
    val CARD_VALUES = "23456789TJQKA".toArray
    val HAND_RANKS = Array (
      "HighCard",
      "OnePair",
      "TwoPair",
      "ThreeKind",
      "Straight",
      "Flush",
      "FullHouse",
      "FourKind",
      "StraightFlush")

    val values = for (c <- cards) yield (CARD_VALUES.indexOf(c(0)) + 1)
    val suits = for (c <- cards) yield c(1)

    lazy val rank: Int = HAND_RANKS.indexOf {
      val distinct = values.toSet.size
      lazy val longest = (
        for ((_, group) <- values.groupBy(identity)) yield group.length
      ).max

      distinct match {
        case 2 => if (longest == 4) "FourKind" else "FullHouse"
        case 3 => if (longest == 3) "ThreeKind" else "TwoPair"
        case 4 => "OnePair"
        case 5 => {
          val straight = values.max - values.min == 4
          val flush = suits.toSet.size == 1
          if (straight && flush) "StraightFlush"
          else if (straight) "Straight"
          else if (flush) "flush"
          else "HighCard"
        }
      }
    }

    override def compare(that: Hand): Int = {
      def tiebreakers(values: Array[Int]) = {
        def groupScore(value: Int, group: Array[Int]) =
          value * math.pow(CARD_VALUES.length+1, group.length-1).toInt

        (values groupBy identity map tupled(groupScore)).toArray.sorted.reverse
      }
      if (this.rank == that.rank) {
        val (thisTiebreaker, thatTiebreaker) =
          (tiebreakers(this.values) zip tiebreakers(that.values)
            dropWhile tupled(_ == _)).head

        thisTiebreaker - thatTiebreaker
      } else {
        this.rank - that.rank
      }
    }
  }

  def main(args: Array[String]) {
    val filename = if (args.nonEmpty) args(0) else "p054.txt"

    val t0 = System.nanoTime / 1000000
    val ans = (io.Source.fromFile(filename).getLines map (_.split(' '))
      map (cards => (new Hand(cards.slice(0,5)), new Hand(cards.slice(5,10))))
      filter tupled(_ > _)).length
    val t1 = System.nanoTime / 1000000

    println(t1 - t0)

    println(ans)
  }
}
